pyo3-macros-backend/src/deprecations.rs

```diff
@@ -45,3 +48,50 @@ impl<'ctx> ToTokens for Deprecations<'ctx> {
         }
     }
 }
+
+pub(crate) fn deprecate_trailing_option_default(spec: &FnSpec<'_>) -> TokenStream {
+    if spec.signature.attribute.is_none()
+        && spec.signature.arguments.iter().any(|arg| {
+            if let FnArg::Regular(arg) = arg {
+                arg.option_wrapped_type.is_some()
+            } else {
+                false
+            }
```

Comment:
Is any arg having `option_wrapped_type` strict enough of a check? Does it need to be something like 

1. filter to regular args
2. check if the last regular arg is an option

?

Reply:
I think it is enough, because we currently reject required arguments after `Option`s without a signature:
```rust
#[pyfunction]
fn foo(_i: u32, _any: Option<i32>, _foo: String) {}
```
Throws:
```
error: required arguments after an `Option<_>` argument are ambiguous
       = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
  --> src/lib.rs:82:11
   |
82 |     _foo: String,
   |           ^^^^^^
```
So if we find `Option` args without `signature` attribute they have to be trailing. Do you agree?

(But changing it to your proposal also seems relatively straight forward, so I can also do that if you prefer)

Reply:
Ah, good point.

I wonder, does it make sense to remove that error as a follow-up to this PR? After all, in the model we're heading towards that `foo` is no longer ambiguous and is simply just 3 arguments without defaults. There's also no trailing `Option<_>` arguments which added the special case before.

That's probably a thorny enough change that I'm happy if we leave the above as-is for now and change in the follow-up...

Reply:
> I wonder, does it make sense to remove that error as a follow-up to this PR? After all, in the model we're heading towards that `foo` is no longer ambiguous and is simply just 3 arguments without defaults.

Definitely, but I think we can only do that after the migration period, right? So If we include this in 0.22, we should require the signature attribute in 0.23 therefore can remove this ambiguity error, then in 0.24 we can remove the signature requirement again and have the new default.

Reply:
Ah yes, of course. If we remove the error now, then users who ignore this deprecation warning and continue to use implicit trailing optional arguments may accidentally refactor them away (that's what the "ambiguity" error is designed to prevent).